src/ccl_core.c

```diff
@@ -386,47 +386,80 @@ ccl_parameters ccl_parameters_create(
     return params;
   }
   
-  // Decide how to split sum of neutrino masses between 3 neutrinos. See the 
-  // CCL note for how we get these expressions for the neutrino masses in 
-  // normal and inverted hierarchy.
+  // Decide how to split sum of neutrino masses between 3 neutrinos. We use
+  // a Newton's rule numerical solution (thanks M. Jarvis).
+  
   if (mnu_type==ccl_mnu_sum){
 	  // Normal hierarchy
+	  
 	  mnu_in = malloc(3*sizeof(double));
-	  double nfac = -6.*DELTAM12_sq + 12.*DELTAM13_sq_pos + 4.*mnusum*mnusum;
 	  
-	  mnu_in[0] = 2./3. * mnusum - 1./6. * pow(nfac, 0.5) 
-	            - 0.25 * DELTAM12_sq / (2./3.* mnusum - 1./6.*pow(nfac, 0.5));
-	  mnu_in[1] = 2./3.* mnusum - 1./6. * pow(nfac, 0.5) 
-	            + 0.25 * DELTAM12_sq / (2./3.* mnusum - 1./6. * pow(nfac, 0.5));
-	  mnu_in[2] = -1./3. * mnusum + 1./3 * pow(nfac, 0.5); 
+	  // Check if the sum is zero
+	  if (*mnu<1e-15){
+		  mnu_in[0] = 0.;
+		  mnu_in[1] = 0.;
+		  mnu_in[2] = 0.;
+	  } else{
+	  
+	      mnu_in[0] = 0.; // This is a starting guess.
 	  
-	  if (mnu_in[0]<0 || mnu_in[1]<0 || mnu_in[2]<0){
-	    // The user has provided a sum that is below the physical limit.
-	    if (params.sum_nu_masses < 1e-14){
-			mnu_in[0] = 0.; mnu_in[1] = 0.; mnu_in[2] = 0.;
-		}else{
-			*status = CCL_ERROR_MNU_UNPHYSICAL;
-	    }
+	      double sum_check;
+	      // Check that sum is consistent
+	      mnu_in[1] = sqrt(DELTAM12_sq);
+	      mnu_in[2] = sqrt(DELTAM13_sq_pos);
+	      sum_check = mnu_in[0] + mnu_in[1] + mnu_in[2];
+	      if (ccl_mnu_sum < sum_check){
+		      *status = CCL_ERROR_MNU_UNPHYSICAL;
+          }
+      
+          double dsdm1;
+          // This is the Newton's method
+          while (fabs(*mnu - sum_check) > 1e-15){
```

Comment:
Would it make sense to add a counter and exit condition in case the target accuracy doesn't get reached for some reason? E.g., have it less than `ccl_gsl->N_ITERATION`.

Reply:
I don't think that is necessary.  This particular function is extremely smooth for all values of sum_check, and there is no regime where the derivative could possibly shoot off into infinity.  As a result, it rarely takes more than 4 iterations to reach machine precision.

Reply:
Actually, I guess the one thing that could possibly cause problems is if mnu >> 1.  It would still quickly converge to a relative accuracy of machine precision, but possibly not absolute accuracy.

So if you are concerned about that (not physical, but maybe chains could get there) case, then the easy fix would be to switch the check to compare agains `1e-15 * (*mnu)` to make it a relative accuracy test.